File: apps/cli/src/commands/account.rs
use super::detect;
use crate::cli::{AccountImportArgs, Cli};
use crate::config::{self, AccountProfile};
use crate::error::{AppError, AppResult};
use crate::output::{self, AccountMeta, Meta};
use serde::Serialize;
use std::process::Command;

#[derive(Debug, Serialize)]
struct AccountList {
    active: String,
    accounts: Vec<AccountMeta>,
}

#[derive(Debug, Serialize)]
struct Balance {
    selector: String,
    wei: Option<String>,
}

#[derive(Debug, Serialize)]
struct SignerList {
    active: String,
    signers: Vec<String>,
}

#[derive(Debug, Serialize)]
struct AccountAction {
    action: String,
    name: String,
    active: String,
    account: AccountMeta,
    config_path: String,
}

pub fn list(cli: &Cli) -> AppResult<()> {
    let account = detect::active_account(cli)?;
    let config = config::load()?;
    let mut accounts = vec![config::account_meta_from_selector(&config, "anvil0")?];
    if std::env::var("ETH_PRIVATE_KEY").is_ok() {
        accounts.push(config::account_meta_from_selector(&config, "env")?);
    }
    for name in config.accounts.keys() {
        accounts.push(config::account_meta_from_selector(&config, name)?);
    }
    let data = AccountList {
        active: account.name.clone(),
        accounts,
    };

    if cli.json {
        let mut meta = Meta::new("account list");
        meta.account = Some(account);
        output::print_json(data, meta)
    } else {
        println!("Active account: {}", data.active);
        for account in data.accounts {
            println!(
                "  {} {} signer={}",
                account.name,
                account.address.as_deref().unwrap_or("address unknown"),
                account.signer
            );
        }
        Ok(())
    }
}

pub fn use_account(cli: &Cli, selector: &str) -> AppResult<()> {
    let mut config = config::load()?;
    let account = config::account_meta_from_selector(&config, selector)?;
    if account.signer == "selected" {
        return Err(AppError::user(
            "account_not_found",
            format!("Account profile `{selector}` does not exist."),
            Some(
                "Run `consol account list` or import one with `consol account import`.".to_string(),
            ),
        ));
    }
    config.active_account = Some(selector.to_string());
    config::save(&config)?;
    print_action(
        cli,
        AccountAction {
            action: "selected".to_string(),
            name: selector.to_string(),
            active: selector.to_string(),
            account,
            config_path: config::config_path().display().to_string(),
        },
    )
}

pub fn import(cli: &Cli, args: &AccountImportArgs) -> AppResult<()> {
    if args.name == "anvil0" || args.name == "env" {
        return Err(AppError::user(
            "account_reserved",
            format!("`{}` is a built-in account name.", args.name),
            Some("Use a different account profile name.".to_string()),
        ));
    }

    let address = std::env::var(&args.private_key_env)
        .ok()
        .and_then(|key| config::private_key_address(&key));
    let mut config = config::load()?;
    config.accounts.insert(
        args.name.clone(),
        AccountProfile {
            address: address.clone(),
            private_key_env: args.private_key_env.clone(),
            signer: Some("env-private-key".to_string()),
        },
    );
    config::save(&config)?;
    let account = AccountMeta {
        name: args.name.clone(),
        address,
        signer: "env-private-key".to_string(),
    };
    print_action(
        cli,
        AccountAction {
            action: "imported".to_string(),
            name: args.name.clone(),
            active: config
                .active_account
                .clone()
                .unwrap_or_else(|| "anvil0".to_string()),
            account,
            config_path: config::config_path().display().to_string(),
        },
    )
}

pub fn balance(cli: &Cli, selector: Option<&str>) -> AppResult<()> {
    let account = detect::active_account(cli)?;
    let network = detect::active_network(cli)?;
    let selector = selector
        .map(ToOwned::to_owned)
        .or_else(|| account.address.clone())
        .unwrap_or_else(|| account.name.clone());
    let data = Balance {
        wei: balance_wei(&selector, &network.rpc_url),
        selector,
    };

    if cli.json {
        let mut meta = Meta::new("account balance");
        meta.account = Some(account);
        meta.network = Some(network);
        output::print_json(data, meta)
    } else {
        println!(
            "Balance {}: {} wei",
            data.selector,
            data.wei.as_deref().unwrap_or("unknown")
        );
        Ok(())
    }
}

pub fn signer_list(cli: &Cli) -> AppResult<()> {
    let account = detect::active_account(cli)?;
    let data = SignerList {
        active: account.signer.clone(),
        signers: vec![account.signer.clone()],
    };

    if cli.json {
        let mut meta = Meta::new("signer list");
        meta.account = Some(account);
        output::print_json(data, meta)
    } else {
        println!("Active signer: {}", data.active);
        for signer in data.signers {
            println!("  {signer}");
        }
        Ok(())
    }
}

fn print_action(cli: &Cli, data: AccountAction) -> AppResult<()> {
    if cli.json {
        let mut meta = Meta::new(format!("account {}", data.action));
        meta.account = Some(data.account.clone());
        output::print_json(data, meta)
    } else {
        println!("account {}: {}", data.action, data.name);
        println!("  active: {}", data.active);
        println!(
            "  address: {}",
            data.account.address.as_deref().unwrap_or("unknown")
        );
        println!("  signer: {}", data.account.signer);
        println!("  config: {}", data.config_path);
        Ok(())
    }
}

pub fn signer_status(cli: &Cli, _name: Option<&str>) -> AppResult<()> {
    let account = detect::active_account(cli)?;
    if cli.json {
        let mut meta = Meta::new("signer status");
        meta.account = Some(account.clone());
        output::print_json(account, meta)
    } else {
        println!("Signer: {}", account.signer);
        println!("  account: {}", account.name);
        println!(
            "  address: {}",
            account.address.as_deref().unwrap_or("unknown")
        );
        Ok(())
    }
}

fn balance_wei(selector: &str, rpc_url: &str) -> Option<String> {
    let output = Command::new("cast")
        .args(["balance", selector, "--rpc-url", rpc_url])
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

File: packages/cli/src/commands/account.ts
import {
  accountMetaFromSelector,
  activeAccountMeta,
  defaultAccountMeta,
  loadConsolConfig,
  ProjectError,
  saveAccountProfile,
  saveActiveAccount,
} from "@consol/core";
import { runCastBalance } from "@consol/foundry";
import { createSuccessEnvelope } from "@consol/protocol";
import type { AccountMeta } from "@consol/protocol";
import type { GlobalArgs } from "../args";
import type { CliEnv, CliResult } from "../main";
import { VERSION } from "../version";
import { resolveCliReadNetworkRuntime } from "./network-runtime";

export type RunAccountCommandInput = {
  readonly globals: GlobalArgs;
  readonly commandArgs: readonly string[];
  readonly cwd: string;
  readonly env: CliEnv;
};

export async function runAccountCommand(input: RunAccountCommandInput): Promise<CliResult> {
  const subcommand = input.commandArgs.find((arg) => arg !== "--json");
  if (subcommand === "import") {
    return accountImport(input);
  }
  if (subcommand === "use") {
    return accountUse(input);
  }
  if (subcommand === "balance") {
    return await accountBalance(input);
  }

  if (subcommand !== "list") {
    return { exitCode: 1, stdout: "", stderr: "Unsupported account command.\n" };
  }

  return accountList(input);
}

function accountList(input: RunAccountCommandInput): CliResult {
  const config = loadConsolConfig(input.env);
  const account = activeAccountForCommand(input);
  const data = {
    active: account.name,
    accounts: [
      defaultAccountMeta(),
      ...(input.env.ETH_PRIVATE_KEY === undefined
        ? []
        : [{ name: "env", address: null, signer: "env-private-key" } satisfies AccountMeta]),
      ...Object.keys(config.accounts)
        .sort((left, right) => left.localeCompare(right))
        .map((name) => accountMetaFromSelector(config, name)),
    ],
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "account list",
        account,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  return { exitCode: 0, stdout: `Active account: ${data.active}\n`, stderr: "" };
}

function accountImport(input: RunAccountCommandInput): CliResult {
  const name = requiredAccountName(input.commandArgs, "import");
  if (name === "anvil0" || name === "env") {
    throw new ProjectError({
      code: "account_reserved",
      message: `\`${name}\` is a built-in account name.`,
      hint: "Use a different account profile name.",
    });
  }
  const privateKeyEnv = requiredPrivateKeyEnv(input.commandArgs);
  const configPath = saveAccountProfile({
    env: input.env,
    name,
    profile: {
      private_key_env: privateKeyEnv,
      signer: "env-private-key",
    },
  });
  const account: AccountMeta = { name, address: null, signer: "env-private-key" };
  const config = loadConsolConfig(input.env);
  return accountAction(input, {
    action: "imported",
    name,
    active: config.active_account ?? "anvil0",
    config_path: configPath,
    account,
  });
}

function accountUse(input: RunAccountCommandInput): CliResult {
  const name = requiredAccountName(input.commandArgs, "use");
  const config = loadConsolConfig(input.env);
  const account = accountMetaFromSelector(config, name);
  if (account.signer === "selected") {
    throw new ProjectError({
      code: "account_not_found",
      message: `Account profile \`${name}\` does not exist.`,
      hint: "Run `consol account list` or import one with `consol account import`.",
    });
  }
  const configPath = saveActiveAccount({ env: input.env, name });
  return accountAction(input, {
    action: "selected",
    name,
    active: name,
    config_path: configPath,
    account,
  });
}

function accountAction(
  input: RunAccountCommandInput,
  data: {
    readonly action: "imported" | "selected";
    readonly name: string;
    readonly active: string;
    readonly config_path: string;
    readonly account: AccountMeta;
  },
): CliResult {
  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: `account ${data.action}`,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }
  return { exitCode: 0, stdout: `account ${data.action}: ${data.name}\n  active: ${data.active}\n`, stderr: "" };
}

async function accountBalance(input: RunAccountCommandInput): Promise<CliResult> {
  const account = activeAccountForCommand(input);
  const network = await resolveCliReadNetworkRuntime({ globals: input.globals, cwd: input.cwd, env: input.env });
  const selector = balanceSelector(input.commandArgs) ?? account.address ?? account.name;
  const result = await runCastBalance({
    cwd: input.cwd,
    env: input.env,
    rpcUrl: network.rpc_url,
    selector,
  });
  const data = {
    selector,
    wei: result.ok ? result.stdout.trim() : null,
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "account balance",
        network: network.meta,
        account,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  return { exitCode: 0, stdout: `Balance ${data.selector}: ${data.wei ?? "unknown"} wei\n`, stderr: "" };
}

function balanceSelector(commandArgs: readonly string[]): string | undefined {
  const index = commandArgs.findIndex((arg) => arg === "balance");
  if (index === -1) {
    return undefined;
  }
  return commandArgs.slice(index + 1).find((arg) => arg !== "--json");
}

function activeAccountForCommand(input: RunAccountCommandInput): AccountMeta {
  if (input.globals.account !== undefined || input.globals.signer !== undefined) {
    const config = loadConsolConfig(input.env);
    const selector = input.globals.account ?? input.globals.signer ?? "anvil0";
    const account = accountMetaFromSelector(config, selector);
    if (account.signer === "selected") {
      throw new ProjectError({
        code: input.globals.account === undefined ? "signer_not_found" : "account_not_found",
        message: `Account profile \`${selector}\` does not exist.`,
        hint: "Run `consol account list` or import one with `consol account import`.",
      });
    }
    return account;
  }
  return activeAccountMeta(input.env);
}

function requiredAccountName(commandArgs: readonly string[], subcommand: "import" | "use"): string {
  const name = commandArgs[1];
  if (name === undefined || name.startsWith("--")) {
    throw new ProjectError({
      code: "account_name_required",
      message: "Missing account profile name.",
      hint: `Use \`consol account ${subcommand} <name>\`.`,
    });
  }
  return name;
}

function requiredPrivateKeyEnv(commandArgs: readonly string[]): string {
  const flagIndex = commandArgs.findIndex((arg) => arg === "--private-key-env");
  const value = flagIndex === -1 ? undefined : commandArgs[flagIndex + 1];
  if (value === undefined || value.startsWith("--")) {
    throw new ProjectError({
      code: "account_import_signer_missing",
      message: "Account import requires a signer source.",
      hint: "Use `--private-key-env <ENV>`.",
    });
  }
  return value;
}
